x-pack/plugins/actions/server/sub_action_framework/basic.ts

```diff
@@ -0,0 +1,160 @@
+import { SubAction } from './types';
+import { ServiceParams } from './types';
+import * as i18n from './translations';
+
+const isObject = (v: unknown): v is Record<string, unknown> => {
+  return typeof v === 'object' && v !== null;
+};
+
+const isAxiosError = (error: unknown): error is AxiosError => (error as AxiosError).isAxiosError;
+
+export abstract class BasicConnector<Config, Secrets> {
+  [k: string]: ((params: unknown) => unknown) | unknown;
+  private axiosInstance: AxiosInstance;
+  private validProtocols: string[] = ['http:', 'https:'];
+  private subActions: Map<string, SubAction> = new Map();
+  private configurationUtilities: ActionsConfigurationUtilities;
+  protected logger: Logger;
+  protected connector: ServiceParams<Config, Secrets>['connector'];
+  protected config: Config;
+  protected secrets: Secrets;
+
+  constructor(params: ServiceParams<Config, Secrets>) {
+    this.connector = params.connector;
+    this.logger = params.logger;
+    this.config = params.config;
+    this.secrets = params.secrets;
+    this.configurationUtilities = params.configurationUtilities;
+    this.axiosInstance = axios.create();
+  }
+
+  private normalizeURL(url: string) {
+    const urlWithoutTrailingSlash = url.endsWith('/') ? url.slice(0, -1) : url;
+    const replaceDoubleSlashesRegex = new RegExp('([^:]/)/+', 'g');
+    return urlWithoutTrailingSlash.replace(replaceDoubleSlashesRegex, '$1');
+  }
+
+  private removeNullOrUndefinedFields(data: unknown | undefined) {
+    if (isObject(data)) {
+      return Object.fromEntries(Object.entries(data).filter(([_, value]) => value != null));
+    }
+
+    return data;
+  }
+
+  private normalizeData(data: unknown | undefined | null) {
+    if (data == null) {
+      return {};
+    }
+
+    return this.removeNullOrUndefinedFields(data);
+  }
+
+  private assertURL(url: string) {
+    try {
+      const parsedUrl = new URL(url);
+
+      if (!parsedUrl.hostname) {
+        throw new Error('URL must contain hostname');
+      }
+
+      if (!this.validProtocols.includes(parsedUrl.protocol)) {
+        throw new Error('Invalid protocol');
+      }
+    } catch (error) {
+      throw new Error(`URL Error: ${error.message}`);
+    }
+  }
+
+  private ensureUriAllowed(url: string) {
+    try {
+      this.configurationUtilities.ensureUriAllowed(url);
+    } catch (allowedListError) {
+      throw new Error(i18n.ALLOWED_HOSTS_ERROR(allowedListError.message));
+    }
+  }
+
+  private getHeaders(headers: Record<string, string>) {
+    return { ...headers, 'Content-Type': 'application/json' };
+  }
+
+  protected registerSubAction(subAction: SubAction) {
+    this.subActions.set(subAction.name, subAction);
+  }
+
+  public getSubActions() {
+    return this.subActions;
+  }
+
+  protected abstract getResponseErrorMessage(error: AxiosError): string;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  protected async request<R = any>({
```

Comment:
Indeed. Some of the code is taken from this method. I did not want to touch the method by adding the needed functionality for the framework because I am not sure how the connectors are being used and I did not want to introduce any bugs. Do you want me to use the existing one?

Reply:
Let's leave it as a follow-up; open a new issue and we can evaluate whether we should do try to get these to use the same function - I suspect you're right we probably need to refactor things a bit - and move them somewhere else.

Reply:
Done! https://github.com/elastic/kibana/issues/132199